ulog.cpp

```diff
@@ -0,0 +1,280 @@
+ *
+ * Copyright (C) 2016  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <mavlink.h>
+#include <string.h>
+
+#include "comm.h"
+#include "log.h"
+#include "ulog.h"
+#include "util.h"
+
+#define TARGET_SYSTEM_ID 1
+#define SYSTEM_ID 2
+
+#define ULOG_HEADER_SIZE 16
+#define ULOG_MAGIC                               \
+    {                                            \
+        0x55, 0x4C, 0x6F, 0x67, 0x01, 0x12, 0x35 \
+    }
+
+#define NO_FIRST_MSG_OFFSET 255
+
+struct _packed_ ulog_msg_header {
+    uint16_t msg_size;
+    uint8_t msg_type;
+};
+
+static bool _ulog_logging_start_timeout(void *data)
+{
+    ULog *ulog = static_cast<ULog *>(data);
+    return ulog->logging_start_timeout();
+}
+
+bool ULog::logging_start_timeout()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_START;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = TARGET_SYSTEM_ID;
+
+    mavlink_msg_command_long_encode(SYSTEM_ID, 0, &msg, &cmd);
+    _endpoint->write_msg(&msg);
+
+    return true;
+}
+
+static bool _on_master_msg_cb(struct buffer *buf, void *data)
+{
+    ULog *ulog = static_cast<ULog *>(data);
+    return ulog->handle_fligh_stack_message(buf);
+}
+
+bool ULog::start(Mainloop *mainloop, Endpoint *master, const char *base_path)
+{
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    char path[1024];
+
+    null_check(mainloop, false);
+    null_check(master, false);
+
+    if (_file) {
+        log_error("ULog already started");
+        return false;
+    }
+
+    snprintf(path, sizeof(path), "%s/%i-%i-%i_%i-%i-%i.ulg", (base_path ? base_path : ""),
+             timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday, timeinfo->tm_hour,
+             timeinfo->tm_min, timeinfo->tm_sec);
+    _file = fopen(path, "wb");
+    if (!_file) {
+        log_error_errno(errno, "Unable to open ULog file(%s): (%m)", path);
+        return false;
+    }
+
+    _endpoint = master;
+    _mainloop = mainloop;
+    _mainloop->set_on_master_msg_callback(_on_master_msg_cb, this);
+    _timeout = _mainloop->timeout_add(MSEC_PER_SEC, _ulog_logging_start_timeout, this);
+    _waiting_header = true;
+    _waiting_first_msg_offset = false;
+    logging_start_timeout();
+
+    return _timeout;
+}
+
+void ULog::stop()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
```

Comment:
Ah, ok. But with `const`, it would live on read-only memory section afair, not on stack.